crates/ra_hir_ty/src/unsafe_validation.rs

```diff
@@ -0,0 +1,138 @@
+//! Provides validations for unsafe code. Currently checks if unsafe functions are missing
+//! unsafe blocks.
+
+use std::sync::Arc;
+
+use hir_def::{DefWithBodyId, FunctionId};
+use hir_expand::diagnostics::DiagnosticSink;
+
+use crate::{
+    db::HirDatabase, diagnostics::MissingUnsafe, lower::CallableDef, ApplicationTy,
+    InferenceResult, Ty, TypeCtor,
+};
+
+use rustc_hash::FxHashSet;
+
+use hir_def::{
+    body::Body,
+    expr::{Expr, ExprId, UnaryOp},
+};
+
+pub struct UnsafeValidator<'a, 'b: 'a> {
```

Comment:
Seems like this (and other stuff) can be `pub(crate)`?

Reply:
These are all used in `rust_hir`, so making them `pub(crate)` would break that crate. This is following the same structure as `ra_hir_ty::expr::ExprValidator`.